#!/usr/bin/env Rscript

#### ENV SET-UP AND ARGUMENTS ####
##### Loading packages silently #####
# Check if one of .libPaths() is writable for the user, and if not, create a writable package library in $HOME (~)
source("~/Rfunctions/check_libPaths.R")
writable_path <- check_libPaths(verbose = FALSE)
# Install/load pacman.
suppressPackageStartupMessages(if (!require(pacman)) {
    install.packages("pacman", lib = writable_path)
})
needed_packs <- c(  "tidyverse", "plyr", "ggplot2", "vegan", "gridExtra", "ggplotify", "ggpubr", "ggtext", "ggforce", "multcompView",
                    "cli", "argparser", "spaa", "apeglm", "this.path", "patchwork", "rio", "abdiv", "picante", "tidytree") 
to_install <- needed_packs[!needed_packs %in% pacman::p_library()]
message("Installing and/or loading required packages...")
if (length(to_install) != 0) {
    pacman::p_install(to_install, character.only = TRUE, try.bioconductor = TRUE, path = writable_path)
}
pacman::p_load(needed_packs, character.only = TRUE)
#github_packs <- c("benjjneb/dada2")
#pacman::p_load_gh(github_packs)

##### Set WD, source functions and variables #####
script_dir <- this.dir() # gets dir where the script is stored
function_dir <- paste0(script_dir, "/functions/")
working_dir <- getinitwd() # gets dir from where the script is launched
source(paste0(function_dir, "/alpha_diversity.R"))

##### Parse arguments with arg_parser #####
cli_h1("Parsing input")
parser <- arg_parser("Check the compatibility of the data (count table, metadata, taxonomy file).")
parser <- add_argument(parser, "counttable", help = "Count table with absolute abundances for each taxon (5_taxonomy.R output)")
parser <- add_argument(parser, "taxonomy", help = "Taxonomy table outputted by 5_taxonomy.R")
parser <- add_argument(parser, "metadata", help = "Metadata file, where Sample_Name column contains sample names exactly as colnames in counttable")
parser <- add_argument(parser, "filtered", help = "Path to filtered_list.RData file, generated by 6.2_filter_transform.R")
parser <- add_argument(parser, "--tree", help = "Path to tree file (nwk) of the ASV sequences. If none supplied, no phylogenetic alpha-diversity will be computed.", default = "None")
parser <- add_argument(parser, "--outdir", help = "Output directory for all the following steps.")
argv <- parse_args(parser)

##### Communicate arguments #####
cli_alert_success("Parsed arguments successfully: ")
cli_li("count table -> {argv$counttable}")
cli_li("filtered_list -> {argv$filtered}")
cli_li("taxonomy -> {argv$taxonomy}")
cli_li("metadata -> {argv$metadata}")
cli_li("tree -> {argv$tree}")

##### Create output directory if necessary #####
outdir <- file.path(argv$outdir)
if (!dir.exists(outdir)) {
    dir.create(outdir, recursive = TRUE)
    cli_alert_info("Output directory {outdir} doesn't exist, creating it.")
} else {
    cli_alert_info("Output directory {outdir} already exists.")
}
##### Load data #####
count_table <- rio::import(argv$counttable) %>% column_to_rownames("V1")
taxonomy <- rio::import(argv$taxonomy) %>% column_to_rownames("V1")
metadata <- rio::import(argv$metadata) %>% column_to_rownames("V1")
tree <- tidytree::read.tree(argv$tree)

##### Set which taxons #####
taxons <- c("phylum", "class", "order", "family", "genus", "species")

#### ALPHA-DIVERSITY ####
##### Compute Alpha-diversity index (Shannon) #####
cli_h1("Computing Shannon alpha-diversity...")
alpha_list <- GetAlphaList(metadata, filtered_list, taxons, index = "shannon")
cli_alert_success("Successfully calculated Shannon alpha-diversity index")
cli_alert_info("Writing alpha diversity tables.")
l_ply(taxons, .fun = function(x) write.table(alpha_list[[x]], file = paste0(outdir, "/alpha-diversity_", x, ".csv"), quote = FALSE, sep = ";", row.names = TRUE, col.names = TRUE))

##### Plot Alpha-diversity (Boxplots) #####
cli_h1("Making alpha-diversity boxplots...")
alpha_plots <- AlphaPlotList(alpha_list, taxons, myColors, index = "shannon")
multi_page <- marrangeGrob(grobs = alpha_plots, ncol = 2, nrow = 2)
ggsave(multi_page, filename = paste0(outdir, "/alpha_diversity.pdf"), width = 20, height = 20)
cli_alert_success("Alpha diversity plots saved to {outdir}/alpha_diversity.pdf")

##### Phylogenetic alpha-diversity #####
cli_h1("Computing Faith's Phylogenetic Diversity...")
alpha_pd <- pd(t(count_table[,-c(1:7)]), tree, include.root=F) %>% 
    rownames_to_column("Sample_Name")
alpha_pd <- left_join(alpha_pd, metadata, by = "Sample_Name")

##### Statistical significance #####
## test normality of alpha_pd distribution
genotypes <- unique(metadata$Genotype)
for ( i in 1:length(genotypes) ){
    normality <- shapiro.test(dplyr::filter(alpha_pd, Genotype == genotypes[i])$PD)$p.value
    if (normality > 0.05) {
        cli_alert_success("Distribution of alpha PDs normal, continuing with ANOVA and tukey for statistical significance")
        ## anova and tukey
        anova <- aov(data = alpha_pd, PD ~ Genotype)
        tukey <- TukeyHSD(anova)
        cld <- as.data.frame(multcompLetters4(anova, tukey)$Genotype$Letters) %>% 
            rownames_to_column("Genotype") %>% 
            setNames(nm = c("Genotype", "Group"))
    } else {
        cli_alert_warning("Distribution of alpha PDs NOT normal, continuing with Kruskal-Wallis and tukey for statistical significance")
        ### add kruskal wallis and posthoc Dunn test https://stackoverflow.com/questions/76857140/compact-letter-display-for-kruskal-wallis-dunn-post-hoc-test-on-r
    }
}

## plot
pd_plot <- ggplot(alpha_pd, aes(x = Genotype, y = PD, fill = Genotype)) + 
    geom_boxplot(outlier.shape = 9, outlier.size = 3, outlier.colour = "black", width = .5) + 
    geom_point(aes(shape = Genotype), alpha = .7, size = 3) +
    theme_bw() +
    scale_fill_brewer(palette = "Pastel1") +
    labs(x = "Genotype", title = paste0("Faith's PD (alpha phylogenetic diversity).")) +
    theme(strip.text = element_text(size = 20), legend.title = element_text(size = 16), legend.text = element_text(size = 14),
    axis.title = element_text(size = 20), axis.text = element_text(size = 16), panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
    guides(x = guide_axis(angle = 45))+
    geom_text(data = cld, aes(x = Genotype, y = 55, label = Group))

all_alpha <- alpha_plots$order + pd_plot + plot_layout(guides = 'collect')
